### ISSUE TITLE
Errors for kindle pw4 2018

### ISSUE THREAD

**Issue by chevcheli0s** [trap000d/dropbox-kindle-client#2]
I tried using this script for pw4 2018.
But it crashes
In the log: 

```
 Traceback (most recent call last):
  File "/mnt/us/extensions/dropbox/bin/dbcli.py", line 512, in <module>
    max_x, max_y = screen_size()
  File "/mnt/us/extensions/dropbox/bin/dbcli.py", line 432, in screen_size
    x = 1 + int(l[3]) / (int(l[0]) / 100)
IndexError: list index out of range
```

I have tried hadrcoding max_x and max_y
As a result, the script works, but there are many errors in the log.

``` 
eips: pixel_in_range> (16, 34752) pixel not in range (0..1072, 0..1448)
eips: pixel_in_range> (32, 34752) pixel not in range (0..1072, 0..1448)
eips: pixel_in_range> (48, 34752) pixel not in range (0..1072, 0..1448)
....
eips: pixel_in_range> (16992, 34728) pixel not in range (0..1072, 0..1448)
eips: pixel_in_range> (17008, 34728) pixel not in range (0..1072, 0..1448)
eips: pixel_in_range> (17024, 34728) pixel not in range (0..1072, 0..1448)
eips: pixel_in_range> (17040, 34728) pixel not in range (0..1072, 0..1448)
eips: pixel_in_range> (17056, 34728) pixel not in range (0..1072, 0..1448)
eips: pixel_in_range> (17072, 34728) pixel not in range (0..1072, 0..1448)
eips: pixel_in_range> (17088, 34728) pixel not in range (0..1072, 0..1448)
eips: pixel_in_range> (17104, 34728) pixel not in range (0..1072, 0..1448)
eips: pixel_in_range> (17120, 34728) pixel not in range (0..1072, 0..1448)
eips: pixel_in_range> (17136, 34728) pixel not in range (0..1072, 0..1448)
eips: pixel_in_range> (17152, 34728) pixel not in range (0..1072, 0..1448)
update_to_display: update_mode=PARTIAL, wave_mode=257 inverted=0
swipe feature is not supported in this platfom G2
eips: pixel_in_range> (64, 34728) pixel not in range (0..1072, 0..1448)
eips: pixel_in_range> (80, 34728) pixel not in range (0..1072, 0..1448)
eips: pixel_in_range> (96, 34728) pixel not in range (0..1072, 0..1448)
eips: pixel_in_range> (112, 34728) pixel not in range (0..1072, 0..1448)
update_to_display: update_mode=PARTIAL, wave_mode=257 inverted=0
swipe feature is not supported in this platfom G2
```

**Comment by trap000d** [comment]
Well, seems max screen coordinates of PW4 are changed [or eips syntax?]. Unfortunately I have no PW4 on a hand to test. Moreover, my PW3 has lost jailbreak after accidental upgrade to 5.13.5. so I can't debug all this stuff until manage to "fix" it (the only option I have now is serial jailbreak).
All I see from your logs is that screen size wasn't detected properly. How it worked in the past: I run eips to output a single character out of screen in particular position, got error with real screen size in pixels, parsed it and then calculated size in characters
```
def screen_size():
    cmd = Popen('eips 99 99 " "', shell=True, stdout=PIPE)
    # eips: pixel_in_range> (1600, 2400) pixel not in range (0..1072, 0..1448)
    for line in cmd.stdout:
        l = findall(r'\d+', line)
        x = 1 + int(l[3]) / (int(l[0]) / 100)
        y = int(l[5]) / (int(l[1]) / 100)
        return x, y

```
E.g. for horizontal axis: 99th row is on position of 1600 pixels, whereas maximum size is 1072, here the character width is 1600/(99+1)=16px, and screen width in chars (max_x) is 1072/16=67

The "right" way to hardcode ```max_x``` and ```max_y``` is the following (you have to know proper values, though):
```
def screen_size():
        x = 67
        y = 60
        return x, y
```

Anyway, if by a chance you have terminal on your Kindle, could you run 
```
eips 99 99 " "
eips 199 199 " "
```
and send me output?

**Comment by chevcheli0s** [comment]
kterm:
![IMG_20211108_092449](https://user-images.githubusercontent.com/48658076/140695549-64694f5e-01a8-4d9e-8d29-4a0de52acfbb.jpg)

**Comment by trap000d** [comment]
OK. Looks like they've added some extra info into error message. So there is a chance my script is parsing it improperly and catching wrong numbers (Regexp: "Now you have two problems."). Unfortunately I can't perform any tests until I unlock my own Kindle. I've just created a branch 'pw4' with hardcoded PW4 screen size. You might check it out and test if it works for you.
```
git pull
git checkout pw4
```

**Comment by chevcheli0s** [comment]
Thanks!
But I tried same solution(deleted func screen_size() and hardcoded vars in main).
I wrote about it in the first post.
Is values of vars x = 67 and y = 60 actual for PW4 with resolution 1072×1448?

**Comment by trap000d** [comment]
Sorry, with no access to device I can only speculate.
```
eips: pixel_in_range> (17152, 34728) pixel not in range (0..1072, 0..1448)
```
There shouldn't be such large numbers like 17152 and 34728: max_x and max_y are small enough (indeed, 67 and 60 for PW3 and PW4), so you might expect to send to eips something in range 16...48 for X axis and 1368...1392 - for Y. Could you either define these variables in wrong place or with wrong indent (this is Python: watch for tabs and spaces in code)?

To trace/debug you can put print here and there and then run script from terminal. I used to run it through SSH when had writing the code and found it very convenient (just disable sleep mode). Running script on a regular computer makes pretty little sense for obvious reasons :)

P.S. I might take a look at the code if ever will have intention to take my Kindle apart and find time for it.

**Comment by trap000d** [comment]
Following up the issue. Few days ago I've managed to get full access to internals of my PW3 without hardware modifications :)
And restored access to my DropBox account (I haven't used it since 2018 or 19)

I have to say - code still works. I can't reproduce the issue. According to the report, it seems you've modified the code.

```Python
Traceback (most recent call last):
  File "/mnt/us/extensions/dropbox/bin/dbcli.py", line 512, in <module>
    max_x, max_y = screen_size()
  File "/mnt/us/extensions/dropbox/bin/dbcli.py", line 432, in screen_size
    x = 1 + int(l[3]) / (int(l[0]) / 100)
IndexError: list index out of range
```
https://github.com/trap000d/dropbox-kindle-client/blob/30ee0f4c5b7cff4db6504c2e2b2bbde880753251/KUAL/dropbox/bin/dbcli.py#L508
https://github.com/trap000d/dropbox-kindle-client/blob/30ee0f4c5b7cff4db6504c2e2b2bbde880753251/KUAL/dropbox/bin/dbcli.py#L428

As I don't know what was changed, I can't diagnose it.
So I'm closing it.
